Develop

Etschbeijer opened the pull request:
Merge the different features into the master branch

Etschbeijer committed:
* Created basic project structure
* Installed required nuget packages

Etschbeijer committed:
Add basic classes and methods for gibss sampling
Different projects for different matrixes and functions

Etschbeijer committed:
Merge pull request #2 from Etschbeijer/feature-ImplementationOfBasicData

* Created basic project structure for gibbs sampling

Etschbeijer committed:
Add positive unit tests for different, not heuristic functions

Etschbeijer committed:
Remove unnecessary collections

Etschbeijer committed:
Add heuristic unit test for finding motif

Etschbeijer committed:
Merge pull request #4 from Etschbeijer/Feature-AddUnitTests

Feature add unit tests

Freymaurer commented on GibbsSampling.Application.Tests/Fixtures/Fixture.fs:
```diff
@@ -0,0 +1,91 @@
+type Fixture() =
+    
+    member this.Pseudocount with get () = 1.
+    
```
I think in f# this syntax can improved by using:

- a module with static parameters
- a static class (`type xxx =` without the `()`) with static members

```fsharp
type Fixture =
    static member Pseudocount = 1.
```

This shows clear intent, that these are read only values without initializing a mutable class first.

Freymaurer replied:
In addition it improves access by changing `Fixture().Pseudocount` to `Fixture.Pseudocount`. No need to init a class on every parameter access.

Etschbeijer replied:
fixed

Freymaurer replied:
still `type Fixture() =` 😅

Freymaurer replied:
but with static members access seems the same, so you can leave it or update it as you want

caroott commented on GibbsSampling.Library/CompositeVector/CompositeVectorTypes.fs:
BaseMatrix<'a, 'value when 'a :> IBioItem>internal(matrix:'value [,])

do you think the Matrix type from FSharp.Stats could be used here?

caroott replied:
and why do you have the guard clause for casting to IBioItem and not just IBioItem?

Etschbeijer replied:
Issue https://github.com/Etschbeijer/GibbsSampling2.0/issues/9 has been created to check for the use of the matrix type

Freymaurer commented on GibbsSampling.Library/CompositeVector/CompositeVectorTypes.fs:
```diff
@@ -0,0 +1,73 @@
+module Types =
+    
+    /// Matrix with fixed positions for nucleotides and amino acids.
+    type BaseMatrix<'a, 'value when 'a :> IBioItem>internal(matrix:'value [,]) =
```
Here you are always working on the original matrix given into the BaseMatrix type.

It might be beneficial store a copy of it inside the type to increase robustness against mutable changes. But this depends on the performance of course.

```fsharp
type BaseMatrix (matrix) =
    member val _matrix = matrix.Copy() with get, set
```

Here copy represents any logic necessary to copy the mutable object. then work on `_matrix` instead of `matrix`

Etschbeijer replied:
fixed

Freymaurer commented on GibbsSampling.Library/CompositeVector/CompositeVectorTypes.fs:
```diff
@@ -0,0 +1,73 @@
+module Types =
+    
+    /// Matrix with fixed positions for nucleotides and amino acids.
+    type BaseMatrix<'a, 'value when 'a :> IBioItem>internal(matrix:'value [,]) =
```
This type syntax `type BaseMatrix<'a, 'value when 'a :> IBioItem>internal(matrix:'value [,])` looks extremely complex to me. If possible dump it down to increase readability.

Etschbeijer replied:
Created new issue:
https://github.com/Etschbeijer/GibbsSampling2.0/issues/11

caroott commented on GibbsSampling.Library/GibbsSampling.Library.fsproj:
BioFSharp Version 1.2.0 is 4 years old. 2.0.0-preview.3 is the current version. Additionally, you don't pin versions with the "Include="BioFSharp" Version="1.2.0"" Syntax. If you want to pin a version you have to write "Include="BioFSharp" Version="[1.2.0]"".

Etschbeijer replied:
fixed, updated the version

Freymaurer commented on GibbsSampling.Library/MotifSampler/MotifSamplerFunctions.fs:
```diff
@@ -0,0 +1,334 @@
+    /// Calculates all possible segment combination of m segments, which do not overlap in the given width 
+    /// and gives you the probability for it.
+    let calculatePWMsForSegmentCombinations (cutOff:float) (width:int) (m:int) (set:list<float*int>) =
+        let rec loop prob positions size set = 
```
readability can be improved here:

- As far as i understand this function it can only ever return a sequence with a single element or none element. If this is true, remove all sequence logic and return a option, on which you can use List.singleton or List.empty in the end.
- Use guard cases in your pattern matching to reduce if-logic tree
    `| n, x::xs when n > 0`

Etschbeijer replied:
Issue erstellt: https://github.com/Etschbeijer/GibbsSampling2.0/issues/10

caroott commented on GibbsSampling.Library/MotifSampler/MotifSamplerFunctions.fs:
```diff
@@ -0,0 +1,334 @@
+    /// The amount of combinations is given by the amount of motifs.
+    let calculateNormalizedSegmentScores (cutOff:float) (motifAmount:int) (motifLength:int) (source:BioArray.BioArray<#IBioItem>) (pcv:ProbabilityCompositeVector) (pwMatrix:PositionWeightMatrix) =
+        let segments =
+            let rec loop n acc =
```
List.foldi?

Freymaurer commented on GibbsSampling.Library/PositionMatrix/PositionMatrixFunctions.fs:
```diff
@@ -0,0 +1,137 @@
+    /// Get an integer which is between 0 and the length of the sequence - segmentLength
+    let internal getRandomNumberInSequence (segmentLength:int) (source:'a[]) =
+        let rnd = System.Random()
+        Array.init 1 (fun _ -> rnd.Next(0, source.Length-segmentLength+1))
```
Can remove Array.init and Array.head

Etschbeijer replied:
fixed

Freymaurer commented on GibbsSampling.Application.Tests/Tests/SuccessfullTests.fs:
interesting file name

Etschbeijer replied:
In the corresponding unsuccessful file you would check for exception handling.

Freymaurer commented on GibbsSampling/Tests/SuccessfullTests.fs:
empty file?

Etschbeijer replied:
fixed

Freymaurer commented on GibbsSampling.Library/SiteSampler/SiteSamplerFunctions.fs:
```diff
@@ -0,0 +1,494 @@
+                else loop 0 acc (Array.copy acc)
+            else
+                let unChosenStartPositions =
+                    Array.append acc.[0..randomSourceNumber.[n]-1] acc.[randomSourceNumber.[n]+1..]
```
Array.foldi or [|for loop|]

Freymaurer replied:
Array.append is very slow

Etschbeijer replied:
fixed

Freymaurer reviewed (changes requested):
I have to admit i skipped looking into the complex functions in detail.

caroott commented on GibbsSampling.Library/SiteSampler/SiteSamplerFunctions.fs:
```diff
@@ -0,0 +1,494 @@
+                     else acc                    
+                    )
+                    bestmotive
+        loop 0 (Array.copy startPositions) (Array.copy startPositions)
```
why copy?

Etschbeijer replied:
fixed

caroott commented on GibbsSampling.Library/PositionMatrix/PositionMatrixFunctions.fs:
```diff
@@ -0,0 +1,137 @@
+    let createPFMOf (source:BioArray.BioArray<#IBioItem>) =
+        let positionFrequencyMatrix = new PositionFrequencyMatrix(source.Length)
+        source
+        |> Array.fold (fun (row, cm) column -> row + 1, increaseInPlacePFM row column cm) (0, positionFrequencyMatrix) |> ignore
```
why do you ignore the result of the fold? if you use an in place function, can't you use Array.iteri (i for row increments)?

caroott replied:
Also, your in place update returns the changed matrix instead of unit, so you could just return the result and not ignore it

Etschbeijer replied:
fixed

Etschbeijer committed:
Update BioFasharp to preview 3 version
Remove empty file
Replace Array.Append with helper function
Remove Array.Copy where not needed

Etschbeijer committed:
Adapt if-else indentation to improve readability
Replace member of fixtures with static members to improve usability

Etschbeijer committed:
increase robusstnusness against value change of base matrix input matrix

Etschbeijer committed:
Merge pull request #7 from Etschbeijer/refactoring-IncludeReviewSuggestions

Refactoring with review suggestions

Freymaurer commented on GibbsSampling.Library/CompositeVector/CompositeVectorFunctions.fs:
```diff
@@ -31,7 +32,7 @@ module Functions =
     let createFCVWithout (motiveLength:int) (position:int) (resSource:BioArray.BioArray<#IBioItem>) =
         let backGroundCounts = new FrequencyCompositeVector()
-        Array.append resSource.[0..(position-1)] resSource.[(position+motiveLength)..]
+        mergeArrays resSource.[0..(position-1)] resSource.[(position+motiveLength)..]
```
what is this meant to do? append, just joins two arrays, right?

Etschbeijer replied:
Yes but we exclude the motif, based on its start position and length ->
We split the array arround the motif (position + length) and merge it again.

Freymaurer replied:
This?

````fsharp
let array1 = [|0..100|]

// mapi approach
let test (motiveStart: int) (motiveLength: int) (arr2: 'b array) =
    arr2
    |> Array.mapi (fun i (x: 'b) -> 
      let motiveEnd = motiveStart+motiveLength
      if i >= motiveStart && i < motiveEnd then
        None
      else
        Some x
    )
    |> Array.choose id

// fold approach
let test2 (motiveStart: int) (motiveLength: int) (arr2: 'b array) =
    arr2
    |> Array.fold (fun (acc: 'b list, i: int) (x: 'b) -> 
      let motiveEnd = motiveStart+motiveLength
      if i >= motiveStart && i < motiveEnd then
        (acc, i+1)
      else
        (x::acc, i+1)
    ) ([],0)
    |> fst
    |> List.rev
    |> List.toArray

```

not sure which is more optimized under the hood.

Etschbeijer committed:
fix merge arrays function, no more swapping of order
